### PR TITLE
8338535: GenShen: some style improvements to source code implementation

### PR DESCRIPTION
**Pull request by kdnilsen**
1. Hide union data structure behind accessor methods for ShenandoahHeuristics::RegionData._u
2. Remove unhelpful comment from shenandoahFullGC.cpp
3. Extraneous white space

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8338535](https://bugs.openjdk.org/browse/JDK-8338535): GenShen: some style improvements to source code implementation (**Bug** - P4)

### Reviewers
 * [William Kemper](https://openjdk.org/census#wkemper) (@earthling-amzn - Committer)
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah.git pull/475/head:pull/475` \
`$ git checkout pull/475`

Update a local copy of the PR: \
`$ git checkout pull/475` \
`$ git pull https://git.openjdk.org/shenandoah.git pull/475/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 475`

View PR using the GUI difftool: \
`$ git pr show -t 475`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/475.diff">https://git.openjdk.org/shenandoah/pull/475.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/shenandoah/pull/475#issuecomment-2297458168)

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.hpp [hunk 1]
```diff
@@ -77,9 +83,55 @@ class ShenandoahHeuristics : public CHeapObj<mtGC> {
...
       size_t _garbage;          // Not used by old-gen heuristics.
       size_t _live_data;        // Only used for old-gen heuristics, which prioritizes retention of _live_data over garbage reclaim
-    } _u;
+    } _region_union;
+#ifdef ASSERT
+    UnionTag _union_tag;
+#endif
   } RegionData;
 
+  static inline void set_RegionData_region_and_garbage(RegionData& region_data, ShenandoahHeapRegion* region, size_t garbage) {
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.hpp:92]: Can these static functions just be instance methods? call sites would then be:
```C++
region_data->set_region_and_garbage(region, garbage);
```

**Review by earthling-amzn — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.hpp (hunk 1 above)
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.hpp:92]: +1

**Review by shipilev — COMMENTED** [review]
Understand the intent: it is good to have guardrail that checks the tags. Stylistic nits:

File: src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.hpp [hunk 2]
```diff
@@ -70,7 +70,7 @@ class ShenandoahAdaptiveHeuristics : public ShenandoahHeuristics {
   virtual ~ShenandoahAdaptiveHeuristics();
 
   virtual void choose_collection_set_from_regiondata(ShenandoahCollectionSet* cset,
-                                                     RegionData* data, size_t size,
+                                                     RegionData data[], size_t size,
```
[the remark this answers is not in the document]
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.hpp:73]: Will revert these changes.

File: src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.hpp (hunk 1 above)
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.hpp:92]: very good suggestions.  it looks much better this way.  will make this change momentarily.

File: src/hotspot/share/gc/shenandoah/shenandoahCollectionSet.hpp [hunk 3]
```diff
@@ -36,7 +36,7 @@ class ShenandoahCollectionSet : public CHeapObj<mtGC> {
   friend class ShenandoahHeap;
   friend class ShenandoahCollectionSetPreselector;
 
-  void establish_preselected(bool *preselected) {
+  void establish_preselected(bool preselected[]) {
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahCollectionSet.hpp:39]: Are we going with `bool* preselected` now?

**Review by earthling-amzn — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/shenandoah/shenandoahCollectionSet.hpp (hunk 3 above)
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahCollectionSet.hpp]: Thanks for finding that.  Fixed.

**Review by ysramakrishna — APPROVED** [review]
LGTM

**Review by earthling-amzn — APPROVED** [review]

**Pull request closed without merging**